C# string interpolation: revised feature description

pkulikov opened the pull request:
This PR is the result of discussions held in #3285. The purpose of this PR is to merge the [keyword article](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/interpolated-strings) into the [token article](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated). Tutorial update and creation of a new How-to article are for other PRs.

During the merging I've also revised the content and coverage of string interpolation in the docs:
- provided links to [composite formatting](https://docs.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting) for more information about formatting;
- simplified part on implicit conversions to highlight how to create culture-specific strings; current text is too detailed and examples make the article too long and are more suitable for the tutorial;
- used interactive code blocks; also tried to reduce amount of examples not to make the article too long;
- checked where string interpolation was mentioned throughout the docs and added links to the reference page
- mentioned string interpolation at some articles related to type formatting from the .NET Standard part of the docs.

pkulikov committed:
Merged and revised string interpolation feature description

BillWagner commented:
I like the direction @pkulikov 

Pinging @rpetrusha who has worked on much of this content as well.

svick commented on samples/snippets/csharp/language-reference/tokens/string-interpolation.cs:
```diff
@@ -0,0 +1,85 @@
+        // Expected output is:
+        // en-US      The cost of this item is $1,000.00.
+        // nl-NL      The cost of this item is € 1.000,00.
+        // Invariant  The cost of this item is ¤1,000.00.
```
I think this a bad example, mostly because I think currency format is a weird feature, especially when combined with current culture. An item cost won't be the same number in all currencies, so I think it doesn't make sense to use it even as an example.

One way to rewrite it would be:

```c#
double solarYearMinutes = 525948.7518;

FormattableString message = $"One solar year is {solarYearMinutes:N3} minutes.";

CultureInfo.CurrentCulture = CultureInfo.GetCultureInfo("cs-CZ");
string messageInCurrentCulture = message.ToString();

CultureInfo specificCulture = CultureInfo.GetCultureInfo("en-IN");
string messageInSpecificCulture = message.ToString(specificCulture);

string messageInInvariantCulture = FormattableString.Invariant(message);

const int CultureOutputAlignment = -10;
Console.WriteLine($"{CultureInfo.CurrentCulture,CultureOutputAlignment} {messageInCurrentCulture}");
Console.WriteLine($"{specificCulture,CultureOutputAlignment} {messageInSpecificCulture}");
Console.WriteLine($"{"Invariant",CultureOutputAlignment} {messageInInvariantCulture}");
// Expected output is:
// cs-CZ      One solar year is 525 948,752 minutes.
// en-IN      One solar year is 5,25,948.752 minutes.
// Invariant  One solar year is 525,948.752 minutes.
```

"One solar year in minutes" is weird number, but it's the best I could come up with that's sufficiently large, has fractional part and isn't currency.

The current culture is not en-US, so that current culture and invariant culture produce different output.

pkulikov replied:
Completely agree with you @svick I've just picked up another constant.
Still thinking how to update the story with dates as well.

pkulikov committed:
Deleted keywords/interpolated-strings.md and fixed doc links

pkulikov committed:
Deleted not used code snippets

pkulikov committed:
Revised string interpolation description in what's new topic

pkulikov committed:
Added links when mentioning string interpolation in the docs

pkulikov committed:
Mention string interpolation in formatting numeric results table

pkulikov committed:
Mention string interpolation in type formatting topics

pkulikov committed:
Addressed feedback; minor clean-up.

pkulikov commented on docs/csharp/language-reference/tokens/interpolated.md:
```diff
@@ -11,21 +12,78 @@ f1_keywords:
+
+[!code-csharp-interactive[specify alignment and format string](../../../../samples/snippets/csharp/language-reference/tokens/string-interpolation.cs#2)]
+
+To include a curly brace ("{" or "}") in the text produced by an interpolated string, use two curly braces, "{{" or "}}". For more information, see [Escaping Braces](../../../standard/base-types/composite-formatting.md#escaping-braces). If an interpolated expression contains characters with special meaning in an interpolated string, such as colon (:) or comma (,), consider delimiting the expressi ...
```
Usually commas separate arguments in method calls, but those are already enclosed. 
As a conditional operator is the one that looks to cause almost all confusion when talking about special characters within an interpolated expression, I've switched to mentioning only it.

pkulikov commented on docs/csharp/language-reference/tokens/interpolated.md:
```diff
@@ -11,21 +12,78 @@ f1_keywords:
+      - A <xref:System.FormattableString.Invariant%2A> method that produces a result string for the <xref:System.Globalization.CultureInfo.InvariantCulture>.
+      - A <xref:System.FormattableString.ToString(System.IFormatProvider)> method that produces a result string for a specified culture.
+
+   Additionaly, inspecting the composite format string and how it renders as a result string might help you protect against an injection attack if you were building a query.
```
I've removed that sentence completely as
- it is not about feature, but about how to use feature in a special way
- it is not related to string interpolation per se, but to string composite formatting.

pkulikov commented on docs/csharp/whats-new/csharp-6.md:
````diff
@@ -351,36 +351,29 @@ can be placed between the curly braces of an interpolated string.
 ```
 
 > [!NOTE]
 > The preceding example is not supported in .NET Core version 1.0.1. It is
````
@rpetrusha @BillWagner is this note still true? What version of .NET Core does support the example?

svick replied:
As far as I can tell:

* No version of .Net Framework or .Net Core supports the weird version of `string.Format` or `FormattableString.GetFormat()` that you replaced. (Maybe it's based on some preview version?)
* .Net Core 1.0 does not support `CultureInfo.CreateSpecificCulture` or `CultureInfo.GetCultureInfo`. If you switch to `new System.Globalization.CultureInfo("de-DE")`, the code above will work on .Net Core 1.0+ and .Net Framework 4.6+ (which is the version that introduced `FormattableString`).

pkulikov replied:
@svick thank you for clarification, I've updated the example.

pkulikov commented:
@svick thank you very much for review; I've addressed all your feedback and left some replies to your comments.
@BillWagner @rpetrusha this PR is ready for review

pkulikov committed:
Updated link in the include file

svick commented on docs/csharp/language-reference/tokens/interpolated.md:
````diff
@@ -22,38 +22,42 @@ ms.author: "ronpet"
 
 ```
-{<interpolated-expression>[,<alignment>][:<format-string>]}
+{<interpolated-expression>[,<alignment>][:<formatString>]}
````
Why is `interpolated-expression` separating words using a hyphen, but `formatString` is using camelCase? I think both should use the same approach, whichever one that is.

pkulikov replied:
I've switched to `formatString` to be consistent with [composite formatting doc](https://docs.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting#format-item-syntax), so everything then should be in camelCase. Thank you for the catch @svick

svick commented on samples/snippets/csharp/language-reference/tokens/string-interpolation.cs:
```diff
@@ -61,13 +61,13 @@ private static void DelimitInterpolatedExpression()
         // <Snippet4>
-        double price = 1000;
-        FormattableString message = $"The cost of this item is {price:C}.";
+        double speedOfLight = 299792.458;
```
I considered speed of light, but not speed of light in km/s. It's definitely a better choice than the constant I suggested.

pkulikov committed:
Updated culture-specific example in what's new; addressed feedback

BillWagner commented:
Thanks for doing all this work. I'm reviewing now.

BillWagner commented on samples/snippets/csharp/language-reference/tokens/string-interpolation.cs:
```diff
@@ -0,0 +1,85 @@
+using System;
```
Can you add the .csproj file so that this is a buildable project?

BillWagner reviewed (changes requested):
Thank you for all the work putting this together.  Overall, it's great and it's a huge improvement.

I had a number of small comments and suggestions.  After you make those changes, you should be able to add the "changes-addressed" label, and I'll give it another look.

pkulikov committed:
Feedback

pkulikov committed:
Added .csproj file

pkulikov commented on docs/csharp/language-reference/tokens/interpolated.md:
```diff
@@ -11,21 +12,81 @@ f1_keywords:
+
+There are three implicit type conversions from an interpolated string:
+
+1. Conversion of an interpolated string to a <xref:System.String> instance that is the result of interpolated string resolution with interpolated expression items being replaced with the properly formatted string representations of their results.
```
@BillWagner I've added only the first part of the sentence as the [string.Format](https://docs.microsoft.com/en-us/dotnet/api/system.string.format?view=netframework-4.7.1#System_String_Format_System_String_System_Object___) doc doesn't mention culture.

pkulikov commented:
@BillWagner thank you for review. I've addressed your feedback. However, I don't see how to add the "changes-addressed" label.
Should I merge the latest master into my branch and resolve the conflict in the redirection file?

BillWagner commented:
@pkulikov Yes, if you can resolve the merge conflicts, we'd appreciate it. In the meantime, I'll review the latest changes.

mairaw commented:
I think only folks with write permissions can add labels @pkulikov. I'll try to resolve the conflict for you, otherwise, we can't merge.

mairaw committed:
Merge branch 'master' into string-interpolation-review

BillWagner reviewed (approved):
This looks great @pkulikov Thanks for making the updates.

I'll approve it, and as soon as the merge conflict is addressed, we'll :shipit:

thanks again for the significant contribution.
